src/utils/UVGenerator.js

```diff
@@ -0,0 +1,173 @@
+// https://github.com/mozilla/Spoke/commit/9701d647020e09d584885bd457eb225e9995c12f
+import XAtlas from '@gfodor/xatlas-web';
+import { Float32BufferAttribute, Uint32BufferAttribute } from 'three';
+
+const AddMeshStatus = {
+	Success: 0,
+	Error: 1,
+	IndexOutOfRange: 2,
+	InvalidIndexCount: 3
+};
+
+export class UVGenerator {
+
+	constructor() {
+
+		this._module = null;
+		this.channel = 1;
+
+	}
+
+	async init() {
+
+		if ( this._module ) return;
+
+		const wasmurl = new URL( '../../node_modules/@gfodor/xatlas-web/dist/xatlas-web.wasm', import.meta.url );
+		this._module = XAtlas( {
+
+			locateFile( path ) {
+
+				if ( path.endsWith( '.wasm' ) ) {
+
+					return wasmurl.toString();
+
+				}
+
+				return path;
+
+			}
+
+		} );
+
+		this._xatlas = await this._module;
+
+	}
+
+	generate( geometries, onProgress = null ) {
+
+		if ( geometries.constructor !== Array ) {
+
+			geometries = [ geometries ];
+
+		}
+
+		const xatlas = this._xatlas;
+		xatlas.createAtlas( );
+
+		const meshInfos = [];
+
+		for ( let i = 0; i < geometries.length; i ++ ) {
+
+			const geometry = geometries[ i ];
+
+			const originalVertexCount = geometry.attributes.position.count;
+			const originalIndexCount = geometry.index.count;
+
+			const meshInfo = xatlas.createMesh( originalVertexCount, originalIndexCount, true, true );
+			xatlas.HEAPU32.set( geometry.index.array, meshInfo.indexOffset / Uint32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.position.array, meshInfo.positionOffset / Float32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.normal.array, meshInfo.normalOffset / Float32Array.BYTES_PER_ELEMENT );
+
+			if ( geometry.attributes.uv ) {
+
+				xatlas.HEAPF32.set( geometry.attributes.uv.array, meshInfo.uvOffset / Float32Array.BYTES_PER_ELEMENT );
+
+			} else {
+
+				const vertCount = geometry.attributes.position.count;
+				geometry.setAttribute( 'uv', new Float32BufferAttribute( new Float32Array( vertCount * 2 ), 2, false ) );
+
+			}
+
+			const statusCode = xatlas.addMesh();
+			if ( statusCode !== AddMeshStatus.Success ) {
+
+				throw new Error( `UVUnwrapper: Error adding mesh. Status code ${ statusCode }` );
+
+			}
+
+			meshInfos.push( meshInfo );
+
+		}
+
+		const params = { padding: 2 };
```

Comment:
That's a good question. I added the padding here to ensure that the edge filler has enough room to have one pixel from each adjacent island. Here are the comments in xatlas regarding the configuration of resolution, padding, and texelsPerUnit:

```
	// Number of pixels to pad charts with.
	uint32_t padding = 0;

	// Unit to texel scale. e.g. a 1x1 quad with texelsPerUnit of 32 will take up approximately 32x32 texels in the atlas.
	// If 0, an estimated value will be calculated to approximately match the given resolution.
	// If resolution is also 0, the estimated value will approximately match a 1024x1024 atlas.
	float texelsPerUnit = 0.0f;

	// If 0, generate a single atlas with texelsPerUnit determining the final resolution.
	// If not 0, and texelsPerUnit is not 0, generate one or more atlases with that exact resolution.
	// If not 0, and texelsPerUnit is 0, texelsPerUnit is estimated to approximately match the resolution.
	uint32_t resolution = 0
```

I explicitly did not set the resolution because I don't want xatlas splitting the atlas into subatlases if it decides it has run out of room: it would add a bunch of complexity to the code I didn't want to deal with. That said, when both resolution and texelsPerUnit are set to zero, I don't actually know how it could be dealing with padding in terms of pixels, since it can't know how big a pixel is. Maybe it assumes it's a 1024x1024 resolution atlas as per the estimator it uses in computing texelsPerUnit.

Reply:
From the comments it sounds like no splitting should happen as long at `texelsPerUnit` is set to 0. Good to know for the future, though.

> Maybe it assumes it's a 1024x1024 resolution atlas as per the estimator it uses in computing texelsPerUnit.

It sounds like 1024x1024 is used. To that end it might make the most sense to parameterize this as a percentage of the UV space in a public option so it's agnostic to resolution. Ie a padding of 2 would be set as `2 / 1024` which can be converted to pixels before passing it into x atlas.